src/Compiler/Utilities/Cancellable.fs

```diff
@@ -0,0 +1,224 @@
+namespace FSharp.Compiler
+
+open System
+open System.Threading
+open Internal.Utilities.Library
+
+[<Sealed>]
+type Cancellable =
+    [<ThreadStatic; DefaultValue>]
+    static val mutable private tokens: CancellationToken list
+
+    static member Tokens
+        with private get () =
+            match box Cancellable.tokens with
+            | Null -> []
+            | _ -> Cancellable.tokens
+        and private set v = Cancellable.tokens <- v
+
+    static member UsingToken(ct) =
+        Cancellable.Tokens <- ct :: Cancellable.Tokens
+
+        { new IDisposable with
```

Comment:
Could this be a single static instance instead of allocating a fresh one each time?
AFAICS it does not close over any value anyway... ?

Reply:
Fixed.